include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,183 @@
+}
+
+TimeControlData::TimeControlData()
+  : is_active(false),
+    start_time(std::numeric_limits<double>::max()),
+    end_time(std::numeric_limits<double>::max()),
+    trigger_interval(-1.0),
+    trigger_every_time_steps(numbers::invalid_timestep)
+{
+}
+
+void
+TimeControlData::print(dealii::ConditionalOStream & pcout, bool const unsteady) const
+{
+  if(unsteady || is_active)
+  {
+    print_parameter(pcout, "TimeControl start time", start_time);
+    print_parameter(pcout, "TimeControl end_time", end_time);
+    if(get_unsteady_evaluation_type(*this) == TimeControlData::EvalType::Interval)
+      print_parameter(pcout, "TimeControl triggers every interval", trigger_interval);
+    if(get_unsteady_evaluation_type(*this) == TimeControlData::EvalType::Timestep)
+      print_parameter(pcout, "TimeControl triggers every timestep", trigger_every_time_steps);
+  }
+}
+
+TimeControl::TimeControl()
+  : EPSILON(1.0e-10),
+    reset_counter(true),
+    counter(0),
+    end_time_reached(false),
+    simulation_was_restarted(false)
+{
+}
+
+void
+TimeControl::setup(TimeControlData const & time_control_data_in)
+{
+  time_control_data = time_control_data_in;
+}
+
+// If setup() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, types::time_step const time_step_number) const
+{
+  if(time_control_data.is_active == false)
+    return false;
+
+  // steady case
+  if(!Utilities::is_unsteady_timestep(time_step_number))
+  {
+    ++counter;
+    return true;
+  }
+
+  // unsteady evaluation
+  AssertThrow(get_unsteady_evaluation_type(time_control_data) != TimeControlData::EvalType::Invalid,
+              dealii::ExcMessage(
+                "Unsteady evaluation not possible since TimeControlData::EvalType::Invalid"));
+
+  if(time < time_control_data.start_time - EPSILON)
+    return false;
+
+  if(time > time_control_data.end_time - EPSILON)
+    end_time_reached = true;
+
+  if(time > time_control_data.end_time + EPSILON)
+    return false;
+
+  bool evaluate = false;
+  if(get_unsteady_evaluation_type(time_control_data) == TimeControlData::EvalType::Timestep)
+  {
+    if((time_step_number - 1) % time_control_data.trigger_every_time_steps == 0)
+    {
+      evaluate = true;
+      ++counter;
+    }
+  }
+  else if(get_unsteady_evaluation_type(time_control_data) == TimeControlData::EvalType::Interval)
+  {
+    // In case of a restart, time might be significantly larger than start_time in the first time
+    // step of the restarted simulation. Hence, we have to reconstruct the counter at the end of the
+    // previous simulation, in order to obtain the correct behavior at the beginning of the
+    // restarted simulation.
+    if(reset_counter)
+    {
+      unsigned int counter_increment =
+        ((time - time_control_data.start_time + EPSILON) / time_control_data.trigger_interval);
+
+      if(counter_increment > 0)
+        simulation_was_restarted = true;
```

Comment:
The name `simulation_was_restarted` is a bit misleading. It can be completely fine that a restarted simulation has counter_increment=0. So we have to be careful with the naming. TimeControl can not provide information about restart or not.

Reply:
In that case: Any idea how to check if a simulation is restarted in PointwiseOutputGenerator?

Reply:
see here https://github.com/exadg/exadg/pull/262/files#r1001530656